geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureSource.scala

```diff
@@ -110,188 +104,115 @@ class GeoMesaFeatureSource(val ds: DataStore with HasGeoMesaStats,
-    visitor match {
-      case v: BoundsVisitor    => v.reset(source.ds.stats.getBounds(source.getSchema, query.getFilter))
-
-      case v: MinVisitor if v.getExpression.isInstanceOf[PropertyName] =>
-        val attribute = v.getExpression.asInstanceOf[PropertyName].getPropertyName
-        minMax(attribute, exact = false).orElse(minMax(attribute, exact = true)) match {
-          case Some((min, _)) => v.setValue(min)
-          case None           => super.accepts(visitor, progress)
-        }
-
-      case v: MaxVisitor if v.getExpression.isInstanceOf[PropertyName] =>
-        val attribute = v.getExpression.asInstanceOf[PropertyName].getPropertyName
-        minMax(attribute, exact = false).orElse(minMax(attribute, exact = true)) match {
-          case Some((_, max)) => v.setValue(max)
-          case None           => super.accepts(visitor, progress)
-        }
-
-      case _ => super.accepts(visitor, progress)
+  trait CachingFeatureSource extends GeoMesaFeatureSource {
+
+    private val featureCache =
+      Caffeine.newBuilder().build(
+        new CacheLoader[Query, SimpleFeatureCollection] {
+          override def load(query: Query): SimpleFeatureCollection =
+            new CachingFeatureCollection(new GeoMesaFeatureCollection(CachingFeatureSource.this, query))
+        })
+
+    abstract override def getFeatures(query: Query): SimpleFeatureCollection = {
+      // geotools bug in Query.hashCode
+      if (query.getStartIndex == null) {
+        query.setStartIndex(0)
+      }
+
+      if (query.getSortBy == null) {
+        featureCache.get(query)
+      } else {
+        // Uses mergesort
+        new SortedSimpleFeatureCollection(featureCache.get(query), query.getSortBy)
+      }
     }
 
-  private def minMax(attribute: String, exact: Boolean): Option[(Any, Any)] =
-    source.ds.stats.getAttributeBounds[Any](source.getSchema, attribute, query.getFilter, exact).map(_.bounds)
-
-  override def reader(): FeatureReader[SimpleFeatureType, SimpleFeature] =
-    source.ds.getFeatureReader(query, Transaction.AUTO_COMMIT)
-
-  override def getBounds: ReferencedEnvelope = source.getBounds(query)
-
-  override def getCount: Int = {
-    if (!open.get) {
-      // once opened the query will already be configured by the query planner,
-      // otherwise do it here
-      source.runner.configureQuery(source.getSchema, query)
-    }
-    source.getCount(query)
+    abstract override def getCount(query: Query): Int = getFeatures(query).size()
   }
 
-  // note: this shouldn't return -1 (as opposed to FeatureSource.getCount), but we still don't return a valid
-  // size unless exact counts are enabled
-  override def size: Int = {
-    val count = getCount
-    if (count < 0) { 0 } else { count }
-  }
-}
-
-object GeoMesaFeatureCollection {
-  private val oneUp = new AtomicLong(0)
-  def nextId: String = s"GeoMesaFeatureCollection-${oneUp.getAndIncrement()}"
-}
-
-trait CachingFeatureSource extends GeoMesaFeatureSource {
+  /*_*/
```

Comment:
What do these comment do?

Reply:
I mean, for some reason, they make me think of Marvin the robot from Hitchhiker's Guide ...

Reply:
haha, nice. they disable intelij's syntax highlighting - this snippet for some reason always shows up as an error if you don't disable it.